cranelift/codegen/src/isa/aarch64/lower_inst.rs

```diff
@@ -65,23 +65,58 @@ pub(crate) fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
         }
         Opcode::Iadd => {
             let rd = get_output_reg(ctx, outputs[0]);
-            let rn = put_input_in_reg(ctx, inputs[0], NarrowValueMode::None);
             let ty = ty.unwrap();
             if !ty.is_vector() {
-                let (rm, negated) = put_input_in_rse_imm12_maybe_negated(
-                    ctx,
-                    inputs[1],
-                    ty_bits(ty),
-                    NarrowValueMode::None,
-                );
-                let alu_op = if !negated {
-                    choose_32_64(ty, ALUOp::Add32, ALUOp::Add64)
+                let mul_insn =
+                    if let Some(mul_insn) = maybe_input_insn(ctx, inputs[1], Opcode::Imul) {
+                        Some((mul_insn, 0))
+                    } else if let Some(mul_insn) = maybe_input_insn(ctx, inputs[0], Opcode::Imul) {
+                        Some((mul_insn, 1))
+                    } else {
+                        None
+                    };
+                // If possible combine mul + add into madd.
+                if let Some((insn, idx)) = mul_insn {
+                    let alu_op = choose_32_64(ty, ALUOp::MAdd32, ALUOp::MAdd64);
+                    let rn_input = InsnInput { insn, input: 0 };
+                    let rm_input = InsnInput { insn, input: 1 };
+
+                    let rn = put_input_in_reg(ctx, rn_input, NarrowValueMode::None);
+                    let rm = put_input_in_reg(ctx, rm_input, NarrowValueMode::None);
+                    let ra = put_input_in_reg(ctx, inputs[idx], NarrowValueMode::None);
+
+                    ctx.emit(Inst::AluRRRR {
+                        alu_op,
+                        rd,
+                        rn,
+                        rm,
+                        ra,
+                    });
+                    // TODO: The below NOP causes the register allocator to behave the same
+                    // as before this optimisation.
+                    //
+                    // Without this NOP, the RA seems to spill (3 more times), which causes
+                    // one function to increase in size, and make bz2.wasm take more instructions
+                    // overall (via `perf stat`).
```

Comment:
@julian-seward1 was going to investigate some weird spilling behavior in the regalloc, and I suspect this may be related. OK to leave this for now, I suppose, since it becomes zero bytes in the output, but hopefully we have a better fix soon!

Reply:
I'd prefer that this NOP was removed before landing.  That it has any effect possibly involves some combination of the RA's spill-cost heuristics and/or isel reaching back for operands over basic block boundaries and/or the possibility that isel has separately created an instruction to do the multiply.  There's no particular reason it would have this effect in other cases, and I'd prefer not to have to reason about it when trying to debug these RA effects -- things are complicated enough without it.